python/pyspark/context.py

```diff
@@ -181,11 +181,6 @@ def _do_init(self, master, appName, sparkHome, pyFiles, environment, batchSize,
         self.appName = self._conf.get("spark.app.name")
         self.sparkHome = self._conf.get("spark.home", None)
 
-        for (k, v) in self._conf.getAll():
```

Comment:
Assuming from https://github.com/apache/spark/commit/615fb649d66b13371927a051d249433d746c5f19, seems like this codes are to make sure `self.environment` are synced with the actual `spark.executorEnv` in JVM side. This isn't needed but seems right to keep it matched. I wouldn't remove this unless there's an issue caused by this.

Reply:
It seems like this is used to fix pyspark's tests, not really needed in actual environment. And this part of code is not to synchronize the configuration of the JVM, but to Ignore it.
We did meet some problems here. The code here would overwrite the environment variables passed by the JVM, it is terrible especially when we optimized the environment variables in the JVM.
I will see if there is any way to solve the test failure in [615fb64](https://github.com/apache/spark/commit/615fb649d66b13371927a051d249433d746c5f19). Even if there is no way, we should add the judgment and won't use it in actual environment.

Reply:
Can you describe which issue you faced? It seems not an issue in the actual enviornment.

Reply:
Also please fill in "Does this PR introduce any user-facing change?" as well.

Reply:
> Can you describe which issue you faced? It seems not an issue in the actual enviornment.

We did meet some problems in actual environment. When the user submits some basic environment configuration, such as LD_LIBRARY_PATH, we will automatically load the current or related environment before JVM starts on yarn or k8s.
For example, We have already put hadoop libraries in cluster and set `LD_LIBRARY_PATH='/usr/native_lib'`.  User use these configuration to load hadoop libraries and python libraries.
```
spark.executorEnv.PYTHONHOME ./python
spark.executorEnv.LD_LIBRARY_PATH $ PYTHONHOME / lib / python2.7 / site-packages: $ LD_LIBRARY_PATH
```
As pyspark always overwrites the jvm environment, the environment, passed into python worker, will be invalidated. LD_LIBRARY_PATH will be configured as `$PYTHONHOME/lib/python2.7/site-packages:$ LD_LIBRARY_PATH`, without any configuration taking effect.

Reply:
It will be great to add a test to demonstrate this problem.

Reply:
> As pyspark always overwrites the jvm environment, the environment, passed into python worker, will be invalidated.

Are you saying you're relying on an environment that was not set by `spark.executorEnv.*`? The correct way to set the environment variables is to use `spark.executorEnv.*` explicitly. It is correct for `spark.executorEnv.*` to overwrite the environment variables.

Reply:
> Are you saying you're relying on an environment that was not set by `spark.executorEnv.*`? 

No. We are just trying to be compatible with the user's environment variable parameters in spark.executorEnv. *.
As for example in the above:
```bash
# these two has same results
spark.executorEnv.LD_LIBRARY_PATH=$PYTHONHOME/lib:$LD_LIBRARY_PATH;
spark.executorEnv.LD_LIBRARY_PATH =./python/lib:/usr/native_lib;
# this is also supported if user doesn't need $LD_LIBRARY_PATH
spark.executorEnv.LD_LIBRARY_PATH =./python/lib
```
We do not prevent users from overriding environment variables, we just try to meet users' needs for using cluster environment variables or their environment variables, like linux.

Reply:
In linux, we can use export LD_LIBRARY_PATH to inherit or overwrite environment. We want to support this in spark as well.
```bash
# we got same result in linux
export LD_LIBRARY_PATH=$PYTHONHOME/lib:$LD_LIBRARY_PATH;
export LD_LIBRARY_PATH=./python/lib:/usr/native_lib;
# we can overwrite environment in linux like this
export LD_LIBRARY_PATH=./python/lib
```

Reply:
The mainly change is in yarn and k8s, not in spark, we should adapt to it.
What I claim here is that, with this code, the environment configuration in JVM and python worker  won't be same, and the configure passed into python worker is awful.
You can try `--conf spark.executorEnv.LD_LIBRARY_PATH=$PYTHONHOME/lib/python2.7/site-packages:$ LD_LIBRARY_PATH`, and find what I mean.

Reply:
> with this code, the environment configuration in JVM and python worker won't be same,

I don't see a clear explanation about it. The code here literally want to keep the JVM and python worker env configs to be the same.